src/v/pandaproxy/rest/handlers.cc

```diff
@@ -116,6 +116,33 @@ get_topics_names(server::request_t rq, server::reply_t rp) {
       });
 }
 
+namespace {
+
+ss::future<ss::sstring> serialize_to_json(
+  json::serialization_format res_fmt, kafka::fetch_response res) {
+    ::json::StringBuffer str_buf;
+    ::json::Writer<::json::StringBuffer> w(str_buf);
+
+    ppj::rjson_serialize_fmt(res_fmt)(w, std::move(res));
+    // TODO Ben: Prevent this linearization
+    return ss::make_ready_future<ss::sstring>(str_buf.GetString());
+}
+
+ss::future<ss::sstring> serialize_with_decompression_units(
+  json::serialization_format res_fmt, kafka::fetch_response res) {
+    return storage::internal::maybe_reserve_decompression_units().then(
```

Comment:
This is a good point, I don't see the compression type present on the `fetch_response` to drive the decision, but since we are working with a topic in this context, maybe there is a way to fetch the topic compression type, I do not see a way to get topic metadata in this function though. Is there any way to access that in handler functions?

Reply:
Whether or not (or how) the data is compressed isn't stored at the topic level, but it's stored with the record batch.  `fetch_response` is autogenerated code but if you follow it all the way down, there are multiple responses from ntps where you can access the record_batch and see if compression is enabled.

Reply:
There is a compression property on the topic here https://github.com/redpanda-data/redpanda/blob/dev/src/v/cluster/types.h#L1547, but I guess you mean something like peeking at the first record batch in the response to figure out the compression? 

This commit was dropped because we are no longer reserving units on decompression.

Reply:
> There is a compression property on the topic here https://github.com/redpanda-data/redpanda/blob/dev/src/v/cluster/types.h#L1547, but I guess you mean something like peeking at the first record batch in the response to figure out the compression?

The topic config is merely a suggestion to the producer, it can decide to use a different or no compression. The actual compression for a record batch is stored in the batch header.

Reply:
Thanks for the explanation, makes sense.

Although acquiring a semaphore unit is no longer needed for this PR, I suppose if we did need to do that, just peeking at one record batch to decide what compression is used for all batches in that topic wouldn't be enough, each batch needs to be examined independently.